### PR TITLE
filter blog by topics on render article route

### PR DESCRIPTION
**Pull request by marjisound**
## What does this change?
This PR adds logic for filtering liveblog on render article route. It'd add implementation for pagination when top mention result is provided. 
- only gets top mentions when key event filter is switched off
- Adds `TopicsLiveBlog` BlockRange for when first page is requested with topics
- Updates `firstPage` logic to have correct pagination when top mentions result is provided
- Updates `findPageWithBlock`  to have correct pagination when top mentions result is provided
- Refactors `firstPage` implementation and extracts the details into separate private functions
- Updates `applyFilters` implementation for when top mentions is provided
- Adds detailed tests for `firstPage` & `findPageWithBlock` for when top mentions is provided
- Updates `DCRFake` in order to enable testing of the the data passed to DCR
- Adds test for `renderArticle` endpoint

> **Note**
**This PR does not cover the `renderJson` route and as a result doesn't add logic for `updates`. This would be done in upcoming PRs.** 

paired with @joecowton1 
## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [x] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [x] On CODE (optional)

### CODE REVIEW

File: article/app/controllers/LiveBlogController.scala
Comment (alinaboghiu) [inline article/app/controllers/LiveBlogController.scala:65]: maybe just getTopMentions?

File: article/app/controllers/LiveBlogController.scala
Comment (alinaboghiu) [inline article/app/controllers/LiveBlogController.scala:371]: maybe a typo? 
```suggestion
    if (filterKeyEvents) None
```

Comment (alinaboghiu) [inline article/app/controllers/LiveBlogController.scala:371]: Minor point but to make this a bit more functional I would actually move this if statement outside this function and only call the function if filterKeyEvents is false

File: article/test/DCRFake.scala
Comment (alinaboghiu) [inline article/test/DCRFake.scala:18]: think the type is pretty clear so maybe this reads a bit easier
```suggestion
  val requestedBlogs = new Queue[PageWithStoryPackage]()
```

File: common/app/model/LiveBlogCurrentPage.scala
Comment (alinaboghiu) [inline common/app/model/LiveBlogCurrentPage.scala:100]: this is very neat and easy to follow

**Review by alinaboghiu — APPROVED** [review]

File: article/app/controllers/LiveBlogController.scala
Comment (marjisound) [reply article/app/controllers/LiveBlogController.scala]: Thanks @alinaboghiu for pointing this out. I think the reason I did it this way was to more easily test it. But I now moved this condition out, and updated the test for it as well. Does it look better?

**Comment by marjisound** [comment]
This PR was fully tested in code so I'll merge it to unblock the rest of the work. But if you have any feedback after the merge, please add them here and I'll make sure to address them in the next PR :)

**Pull request merged**
